### PR TITLE
Add `/api/projects` endpoints

### LINKED ISSUE #37: API Definition [projects GET]
Return a list of ProjectDtos

- [ ] Create ProjectDto entity to match the data used on the leaderboard in each project
- [ ] Fetch Projects from Db
- [ ] Transform to Dtos and return them
- [ ] Update Project Service
- [ ] Move the endpoint to the project folder under /api

### LINKED ISSUE #38: API Definition [ProjectById]
Return Project by id

api/projects/detail/[id]
       - return the DetailView of the project from db

- [ ] Fetch data from db by id
- [ ] Return the Project data
- [ ]  Update the Project Service
- [ ] Move to the api/projects/detail/[id] route

### PR DESCRIPTION
**Pull request by swellander**
## Description
Creates the following endpoints to fetch project data from our db:
- `/api/projects` to return a list of all projects (closes #37)
- `/api/projects/detail/{ID}` to return a single project (closes #38)
- `/api/projects/ids` to return a flat list of all project IDs for our static page generation

### CODE REVIEW

File: packages/nextjs/app/project/[id]/page.tsx [hunk 1]
```diff
@@ -5,10 +5,10 @@ import CustomButton from "~~/components/onchain-impact-dashboard/CustomButton";
 import { Project } from "~~/services/database/schema";
 
 export async function generateStaticParams() {
-  const response = await fetch(`${process.env.NEXT_PUBLIC_API_URL}/stub/projects?limit=100`);
-  const { data }: { data: Project[] } = await response.json();
-  return data.map(item => ({
-    id: item.id,
+  const response = await fetch(`${process.env.NEXT_PUBLIC_API_URL}/stub/project_ids`);
+  const data: Project[] = await response.json();
```
Comment (Franpastoragusti) [inline packages/nextjs/app/project/[id]/page.tsx]: this is a string[] no a project[]

Comment (swellander) [reply packages/nextjs/app/project/[id]/page.tsx]: good catch!

File: packages/firebase/seed.sample.json
Comment (swellander) [inline packages/firebase/seed.sample.json:1]: Updated the seed sample file `projects` with agora projects with OSO id mappings

**Comment by escottalexander** [comment]
I am going to dig in to why the build is still failing. This looks good so nice work!

File: packages/nextjs/app/project/[id]/page.tsx
Comment (escottalexander) [inline packages/nextjs/app/project/[id]/page.tsx:1]: This file won't display for me due to lines 43-56 where `[...data?.github, ...data?.packages]`. If the project doesn't have `github` or `packages` properties then the spread operator is trying to act upon undefined causing this error:
`Error: (intermediate value) is not iterable`.

We should check for the existence of `github` and `packages` before using the spread operator.

**Review by escottalexander — CHANGES_REQUESTED** [review]
One thing with two parts that made me have issues. Almost there. Thanks.

**Comment by escottalexander** [comment]
> I am going to dig in to why the build is still failing. This looks good so nice work!

I might need to reseed the production data for the build to succeed. I will give that a try when you commit the fixes I discussed in my review.

File: packages/nextjs/pages/api/projects/detail/[id].ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+export default async function handler(req: NextApiRequest, res: NextApiResponse) {
+  if (req.method !== "GET") {
+    return res.status(405).json({ message: "Method not allowed." });
+  }
+  const id = req.query.id as string | undefined;
+  if (!id) return res.status(400).json({ message: "Must include project ID" });
+  try {
+    const projectDetail = await getProjectById(id);
+    const projectDetailDto = convertProjectToDto(projectDetail);
```
Comment (escottalexander) [inline packages/nextjs/pages/api/projects/detail/[id].ts:13]: The request from the project page needs more than the dto version. It needs the whole Project.

**Review by escottalexander — CHANGES_REQUESTED** [review]

**Comment by escottalexander** [comment]
@swellander https://github.com/BuidlGuidl/onchain-impact-dashboard/pull/46 has the seed data changes for projects covered. If it gets merged in before this then you can remove your changes to `seed.sample.json` and just get the changes contained in that PR.

**Comment by escottalexander** [comment]
Closing this since all the relevant changes were absorbed by #46. Thanks for your help with this @swellander.

**Pull request closed without merging**